crates/autopilot/src/run_loop.rs

```diff
@@ -632,14 +668,14 @@ impl RunLoop {
         auction_id: i64,
         max_blocks_wait: u64,
     ) -> Result<H256, SettleError> {
-        let current = self.eth.current_block().borrow().number;
+        let block_stream = self.eth.current_block();
+        let current = block_stream.borrow().number;
         let deadline = current.saturating_add(max_blocks_wait);
         tracing::debug!(%current, %deadline, %auction_id, "waiting for tag");
         loop {
-            if self.eth.current_block().borrow().number > deadline {
-                break;
-            }
-
+            // It's impossible for the tx to make it into the current block
+            // so we start out by waiting for the next block to appear.
+            let block = ethrpc::block_stream::next_block(block_stream).await;
```

Comment:
Waiting for the next block appears to get the e2e tests stuck. Not sure why, though. Investigating...

Reply:
Ahh, this is slightly annoying. Looks like the test correctly awaited the block the tx was part of but
```rust
            match self
                .persistence
                .find_tx_hash_by_auction_id(auction_id)
                .await
```
could not find the tx in the DB because it hasn't been processed, yet. :/
And since the testnet only mints blocks on demand we won't try to find the tx on the next block.
That indicates that our settlement indexing also needs to be incorporated into the runloop to ensure that all relevant events have been processed at that point.

Reply:
@sunce86 do you foresee any problems if we adjust the `on_settlement_event_updater` to run in 2 modes:
1. current mode is a background task which picks oldest unprocessed settlement event
2. "manual mode" which allows passing in which tx to process that can be called by the runloop

Are there any assumptions in the updater that events have to be processed in order?

Reply:
Looks like we need to know the `block_number` and `log_index` the settlement with an auction_id which we only have after we already indexed the "raw" settlement event. Very annoying that we don't store the `tx_hash` with every event. That would make this simpler. Let's see what can be done about this.

Reply:
@MartinquaXD I don't forsee any issues with double mode. In manual mode, given the tx hash we can easily get event `block_number` and `log_index` and do the rest the same way it's working right now.

> Very annoying that we don't store the tx_hash with every event. That would make this simpler.

Not sure I understood this one, raw SettlementEvent has (block_number, log_index, hash) or you would like to have a hash somewhere else?

Reply:
> raw SettlementEvent has (block_number, log_index, hash)

True, misread the code.
However, after digging more through the code it looks like there is another issue with the double mode. The event updated needs to have the raw settlement event already indexed before it can work (at least at the moment). Because indexing the raw events happens in another separate background task it could happen that we call the updater manually but it still fails because the raw event was not yet indexed. :/

I sketched out an idea in [06ddc1a](https://github.com/cowprotocol/services/pull/2900/commits/06ddc1a5ecd3f187e75b083a978ce68f765babe3) which should already be as fast as possible but can detect timeouts incorrectly when there are multiple settlements in 1 block and our's is not the first.

Reply:
I don't think we should rely on `OnSettlementEventUpdater` to be on a critical path for runloop though. It's error prone. 

If we definitely need <auction, hash> relation on a critical runloop path so we could optimize runloop submission (to end it before deadline) then I would prefer to extract building <auction, hash> into a separate task (but still do not make it part of event indexing as it contains slow web3 call).

Bottom line, we would have a COMBO condition:
1. new block received
2. events indexed, trigger 3 
3. <auction, hash> relation populated

Now use this COMBO condition in runloop instead of only (1)

Reply:
What you described with your COMBO condition is effectively what my current workaround implements.

The raw event indexing listens to the block stream for updates and can notify the event updater that there is work to be done (this was there before the PR). With my workaround the event updater can also signal to the runloop that it processed something so now the runloop always waits for either the next block or the next processed settlement and reacts accordingly.
Since the whole pipeline relies on message passing there should be minimal latency between indexing a new block and the runloop learning about the fully processed settlement.

It's not really pretty IMO but I think this is effectively what has to happen without requiring a probably big refactoring upfront.